kimchi/src/circuits/polynomials/foreign_mul/foreign_mul_2.rs

```diff
@@ -0,0 +1,75 @@
+/// ForeignMul2 - Foreign field element constraints
+///
+///    Foreign field element F is comprised of three 88-bit limbs L0L1L2
+///
+///    * This circuit gate is used to partially constrain L0 and L1
+///    * Together with two instances of ForeignMul0 the limbs L0 and L1 are fully constrained
+///    * It operates on the Curr row
+///
+/// Column | Curr
+///      0 | 0 (ignored)
+///      1 | plookup L0p4
+///      2 | plookup L0p5
+///      3 | plookup L1p4
+///      4 | plookup L1p5
+///      5 | (ignored)
+///      6 | (ignored)
+///      7 | (ignored)
+///      8 | (ignored)
+///      9 | (ignored)
+///     10 | (ignored)
+///     11 | (ignored)
+///     12 | (ignored)
+///     13 | (ignored)
+///     14 | (ignored)
+use std::marker::PhantomData;
+
+use crate::circuits::{
+    argument::{Argument, ArgumentType},
+    expr::{Column, E},
+    gate::{CurrOrNext, GateType},
+};
+use ark_ff::FftField;
+use CurrOrNext::*;
+
+use super::constraints::sublimb_plookup_constraint;
+
+#[derive(Default)]
+pub struct ForeignMul2<F>(PhantomData<F>);
+
+impl<F> Argument<F> for ForeignMul2<F>
+where
+    F: FftField,
+{
+    const ARGUMENT_TYPE: ArgumentType = ArgumentType::Gate(GateType::ForeignMul2);
+    const CONSTRAINTS: u32 = 5;
+
+    // Constraints for ForeignMul2
+    //   * Operates on Curr row
+    //   * Range constrain sublimbs stored in columns 1 through 4
+    //   * The contents of these cells are the 4 12-bit sublimbs that
+    //     could not be plookup'ed in rows Curr - 3 and Curr - 2
+    //   * Copy constraints are present (elsewhere) to make sure
+    //     these cells are equal to those
+    fn constraints() -> Vec<E<F>> {
+        let w = |i| E::cell(Column::Witness(i), Curr);
+
+        // Row structure
+        //  Column w(i) 0    1       ... 4       5     6     7     ... 14
+        //         Curr limb plookup ... plookup crumb crumb crumb ... crumb
+
+        // Apply range constraints on sublimbs (create 4 12-bit plookup range constraints)
+        // crumbs were constrained by ForeignMul1 circuit gate
+        let mut constraints: Vec<E<F>> =
+            (1..5).map(|i| sublimb_plookup_constraint(&w(i))).collect();
+
+        // Temporary dummy constraint to avoid zero polynomial edge case
+        // and avoid verifier check that all commitments aren't identically zero
```

Comment:
not sure I understand this poseidon gate from the comment :D

Reply:
It might make sense to add some more commentary (`we know the poseidon selector is evaluated and not identically zero, but will be zero whenever this gate is active, so use it as a placeholder`).

Reply:
If memory serves, this was the hack that I suggested. Basically, this constraint is comprised of only the Poseidon selector
```rust
        constraints.push(E::<F>::cell(
            Column::Index(GateType::Poseidon),
            CurrOrNext::Curr,
        ));
```
without further details.
I think @mimoo's comment was referring to the fact that it's confusing to use poseidon here, so I was suggesting that the comment above should be updated to shed some extra light on *why* the poseidon selector appears here as the body of the constraint. Concretely, I was proposing updating the comment to something like
```rust
        // Temporary placeholder constraint.
        // We cannot use any expression that evaluates directly to the zero polynomial
        // here, because the zero group element is not valid as a commitment point, and
        // thus will cause proof generation / verification to fail.
        // Instead, we abuse our knowledge that the poseidon gate is never active at the
        // same time as this gate, that its selector is not identically zero, and that
        // the selector's evaluation is provided in the proof. This allows us to set this
        // constraint as the poseidon selector, which satisfies the requirement of being
        // zero whenever this constraint is active, but importantly also which lets us
        // create a valid (ie. non-zero) commitment.
        // Naturally, this has nothing to do with the poseidon gate; the selector was
        // just an evaluated/committed-to column that matched the requirements to fill
        // this hole until this constraint is implemented.
```

Reply:
This is an interesting `CircuitGate` because it has no constraints other than lookups, which causes it to have a zero polynomial and require the Poseidon workaround.

It seems like we may want another way to handle situations like this without needing this workaround with Poseidon.  @mrmr1993 - are there any other solutions you can think of?

Reply:
You shouldn't need a selector or any constraints for this gate then, only the lookup pattern. I had basically the same scenario when implementing the `Lookup` gate (#399), so that PR might serve as a useful guide.